pkg/network/pasta/pasta.go

```diff
@@ -147,21 +141,18 @@ func (d *parentDriver) ConfigureNetwork(childPID int, stateDir, detachedNetNSPat
 	// `Couldn't open user namespace /proc/51813/ns/user: Permission denied`
 	// Possibly related to AppArmor.
 	cmd := exec.Command(d.binary, opts...)
-	cmd.Stdout = d.logWriter
-	cmd.Stderr = d.logWriter
-	cleanups = append(cleanups, func() error {
-		logrus.Debugf("killing pasta")
-		if cmd.Process != nil {
-			_ = cmd.Process.Kill()
-		}
-		wErr := cmd.Wait()
-		logrus.Debugf("killed pasta: %v", wErr)
-		return nil
-	})
 	logrus.Debugf("Executing %v", cmd.Args)
-	if err := cmd.Start(); err != nil {
+	out, err := cmd.CombinedOutput()
+	if err != nil {
+		exitErr := &exec.ExitError{}
+		if errors.As(err, &exitErr) {
+			return nil, common.Seq(cleanups),
+			       fmt.Errorf("pasta failed with exit code %d:\n%s",
+					  exitErr.ExitCode(), string(out))
+		}
 		return nil, common.Seq(cleanups), fmt.Errorf("executing %v: %w", cmd, err)
 	}
+
```

Comment:
What will happen if the background pasta process dies after several minutes/hours/days?
Can RootlessKit detect that death?

Reply:
It can't, but if it happens (even though we haven't received this kind of "stability" bug reports for quite a long time now), we would like users to report an issue, because anyway context will be lost (TCP connections, at least), rather than hide this by restarting the process or similar.

Reply:
The pasta process can be killed by the kernel mostly due to OOM, and in that case RootlessKit should probably abort, or at least print an error message.

Reply:
It's pretty much impossible because it doesn't allocate memory dynamically, so I don't think it can actually be picked as a candidate by the OOM killer.

Anyway, the network interface will go away in that case, because we close the tap file descriptor, so something like that could actually be added.

I don't think it's actually needed, because the network interface going away is very obvious, but I guess I can implement it if needed (that will take a bit longer though).

Reply:
@AkihiroSuda I'm trying to implement this with the [`LinkSubscribe()`](https://pkg.go.dev/github.com/vishvananda/netlink#LinkSubscribe) method from github.com/vishvananda/netlink.

That part looks trivial, but I don't know where to add this kind of monitor in rootlesskit. I thought there was something like that for slirp4netns' `--exit-fd`, but I can't find it. How would you detect that slirp4netns terminates?

Reply:
Looks like we lack the death detection for slirp4netns too 😅 

Maybe we can use pidfd for kernel >= 5.3.
https://man7.org/linux/man-pages/man2/pidfd_open.2.html

> A PID file descriptor can be monitored using [poll(2)](https://man7.org/linux/man-pages/man2/poll.2.html), [select(2)](https://man7.org/linux/man-pages/man2/select.2.html), and [epoll(7)](https://man7.org/linux/man-pages/man7/epoll.7.html).

If we want to implement this too for old kernel, probably just watch SIGCHLD and see if the PID of `(pasta|slirp4netns)` is still valid

Reply:
> Looks like we lack the death detection for slirp4netns too 😅

I feel better now. :smile:

I think we should leave out PID tracking, because it keeps the interface to pasta very simple (also in terms of AppArmor and SELinux), and we have a valid, solid alternative (netlink monitor on the network interface). Anyway, this is the simple part.

The complicated part for me, whatever we choose as monitor, is where and how to add this to rootlesskit. I really don't know how it works with... threads? Should we spawn a thread somewhere? Is there something like that already in rootlesskit?

Reply:
We don't seem to have any monitoring stuff in the current code base, but it can be probably spawned as a goroutine

Reply:
Thanks for the tip, I'll give it a try, but with very low priority, so it might take a while.